File: android/src/main/kotlin/io/github/harikrishnan_cr/media_compressor/MediaCompressorPlugin.kt
package io.github.harikrishnan_cr.media_compressor

import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import android.os.Handler
import android.os.Looper
import java.io.File

class MediaCompressorPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context
    private lateinit var compressor: NativeCompressor
    private var progressSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        compressor = NativeCompressor(context)

        channel = MethodChannel(binding.binaryMessenger, "native_compressor")
        channel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "native_compressor/progress")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) { progressSink = events }
            override fun onCancel(arguments: Any?) { progressSink = null }
        })
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "compressImage" -> handleCompressImage(call, result)
            "compressVideo" -> handleCompressVideo(call, result)
            "cancel" -> { compressor.cancelVideo(); result.success(null) }
            "release" -> handleRelease(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleCompressImage(call: MethodCall, result: Result) {
        val path = call.argument<String>("path")
        val quality = call.argument<Int>("quality") ?: 80
        val maxWidth = call.argument<Int>("maxWidth")
        val maxHeight = call.argument<Int>("maxHeight")

        if (path == null) { result.error("INVALID_ARGUMENT", "Path is required for image compression", null); return }
        if (quality !in 0..100) { result.error("INVALID_ARGUMENT", "Quality must be between 0 and 100", null); return }
        if (maxWidth != null && maxWidth <= 0) { result.error("INVALID_ARGUMENT", "maxWidth must be greater than 0", null); return }
        if (maxHeight != null && maxHeight <= 0) { result.error("INVALID_ARGUMENT", "maxHeight must be greater than 0", null); return }

        compressor.compressImage(path, quality, maxWidth, maxHeight) { p, e ->
            if (e != null) result.error(e.code, e.message, e.details) else result.success(p)
        }
    }

    private fun handleCompressVideo(call: MethodCall, result: Result) {
        val path = call.argument<String>("path")
        val quality = call.argument<String>("quality") ?: "medium"

        if (path == null) { result.error("INVALID_ARGUMENT", "Path is required for video compression", null); return }
        if (quality.lowercase() !in listOf("low", "medium", "high")) {
            result.error("INVALID_ARGUMENT", "Quality must be one of: low, medium, high", null); return
        }

        val progressCallback: (Float) -> Unit = { progress ->
            Handler(Looper.getMainLooper()).post {
                progressSink?.success(mapOf("progress" to progress, "percentage" to (progress * 100).toInt()))
            }
        }

        compressor.compressVideo(path, quality,
            callback = { p, e -> if (e != null) result.error(e.code, e.message, e.details) else result.success(p) },
            progressCallback = progressCallback)
    }

    // C1 fix: only delete files that live inside the plugin's cache directory.
    private fun handleRelease(call: MethodCall, result: Result) {
        val path = call.argument<String>("path")
        if (!path.isNullOrEmpty()) {
            try {
                val target = File(path).canonicalFile
                val cacheRoot = context.cacheDir.canonicalFile
                if (target.path.startsWith(cacheRoot.path + File.separator) && target.exists()) {
                    target.delete()
                }
            } catch (_: Exception) { /* best-effort */ }
        }
        result.success(null)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        compressor.dispose()
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        progressSink = null
    }
}

File: ios/Classes/MediaCompressorPlugin.swift
import Flutter
import UIKit

public class MediaCompressorPlugin: NSObject, FlutterPlugin {

    private let compressor = NativeCompressor()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "native_compressor",
            binaryMessenger: registrar.messenger()
        )
        let instance = MediaCompressorPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "compressImage": handleCompressImage(call: call, result: result)
        case "compressVideo": handleCompressVideo(call: call, result: result)
        case "cancel": compressor.cancelVideo(); result(nil)
        case "release": handleRelease(call: call, result: result)
        default: result(FlutterMethodNotImplemented)
        }
    }

    private func handleCompressImage(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["path"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                                message: "Path is required for image compression", details: nil)); return
        }
        let quality = args["quality"] as? Int ?? 80
        let maxWidth = args["maxWidth"] as? Int
        let maxHeight = args["maxHeight"] as? Int

        guard quality >= 0 && quality <= 100 else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                                message: "Quality must be between 0 and 100", details: nil)); return
        }
        if let w = maxWidth, w <= 0 {
            result(FlutterError(code: "INVALID_ARGUMENT",
                                message: "maxWidth must be greater than 0", details: nil)); return
        }
        if let h = maxHeight, h <= 0 {
            result(FlutterError(code: "INVALID_ARGUMENT",
                                message: "maxHeight must be greater than 0", details: nil)); return
        }

        compressor.compressImage(imagePath: path, quality: quality,
                                 maxWidth: maxWidth, maxHeight: maxHeight, result: result)
    }

    private func handleCompressVideo(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["path"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                                message: "Path is required for video compression", details: nil)); return
        }
        let quality = args["quality"] as? String ?? "medium"
        guard ["low", "medium", "high"].contains(quality.lowercased()) else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                                message: "Quality must be one of: low, medium, high", details: nil)); return
        }
        compressor.compressVideo(videoPath: path, quality: quality, result: result)
    }

    private func handleRelease(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["path"] as? String, !path.isEmpty else {
            result(nil); return
        }
        // C1 fix: only delete files inside the app's temporary directory.
        let tmp = (NSTemporaryDirectory() as NSString).standardizingPath
        let target = (path as NSString).standardizingPath
        if target.hasPrefix(tmp) {
            try? FileManager.default.removeItem(atPath: target)
        }
        result(nil)
    }
}